common/txmgr/address_state.go

```diff
@@ -218,23 +274,196 @@ func (as *AddressState[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) MoveUn
+	as.Lock()
+	defer as.Unlock()
+
+	tx, ok := as.unconfirmedTxs[txAttempt.TxID]
+	if !ok || tx == nil {
+		return fmt.Errorf("move_unconfirmed_to_confirmed_missing_receipt: no unconfirmed transaction with ID %d", txAttempt.TxID)
+	}
+	if tx.BroadcastAt.Before(broadcastAt) {
+		tx.BroadcastAt = &broadcastAt
+	}
+	tx.State = TxConfirmedMissingReceipt
+	if len(tx.TxAttempts) == 0 {
+		tx.TxAttempts = []txmgrtypes.TxAttempt[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE]{}
+	}
+	txAttempt.State = txmgrtypes.TxAttemptBroadcast
+	tx.TxAttempts = append(tx.TxAttempts, txAttempt)
+
+	as.confirmedMissingReceiptTxs[tx.ID] = tx
+	delete(as.unconfirmedTxs, tx.ID)
+
 	return nil
 }
 
 // MoveInProgressToConfirmedMissingReceipt moves the in-progress transaction to the confirmed missing receipt state.
-func (as *AddressState[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) MoveInProgressToConfirmedMissingReceipt(attempt txmgrtypes.TxAttempt[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE], broadcastAt time.Time) error {
+// If there is no in-progress transaction, an error is returned.
+func (as *AddressState[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) MoveInProgressToConfirmedMissingReceipt(txAttempt txmgrtypes.TxAttempt[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE], broadcastAt time.Time) error {
+	as.Lock()
+	defer as.Unlock()
+
+	tx := as.inprogressTx
+	if tx == nil {
+		return fmt.Errorf("move_in_progress_to_confirmed_missing_receipt: no transaction in progress")
+	}
+	if tx.BroadcastAt.Before(broadcastAt) {
+		tx.BroadcastAt = &broadcastAt
+	}
+	tx.State = TxConfirmedMissingReceipt
+	if len(tx.TxAttempts) == 0 {
+		tx.TxAttempts = []txmgrtypes.TxAttempt[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE]{}
+	}
+	txAttempt.State = txmgrtypes.TxAttemptBroadcast
+	tx.TxAttempts = append(tx.TxAttempts, txAttempt)
+
+	as.confirmedMissingReceiptTxs[tx.ID] = tx
+	as.inprogressTx = nil
+
 	return nil
 }
 
 // MoveConfirmedToUnconfirmed moves the confirmed transaction to the unconfirmed state.
-func (as *AddressState[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) MoveConfirmedToUnconfirmed(attempt txmgrtypes.TxAttempt[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE]) error {
+func (as *AddressState[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) MoveConfirmedToUnconfirmed(txAttempt txmgrtypes.TxAttempt[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE]) error {
+	as.Lock()
+	defer as.Unlock()
+
+	if txAttempt.State != txmgrtypes.TxAttemptBroadcast {
+		return fmt.Errorf("move_confirmed_to_unconfirmed: attempt must be in broadcast state")
+	}
+
+	tx, ok := as.confirmedTxs[txAttempt.TxID]
+	if !ok || tx == nil {
+		return fmt.Errorf("move_confirmed_to_unconfirmed: no confirmed transaction with ID %d", txAttempt.TxID)
+	}
+	tx.State = TxUnconfirmed
+
+	// Delete the receipt from the attempt
+	txAttempt.Receipts = nil
+	// Reset the broadcast information for the attempt
+	txAttempt.State = txmgrtypes.TxAttemptInProgress
+	txAttempt.BroadcastBeforeBlockNum = nil
```

Comment:
To make sure these changes are made to our stored attempt, I think we'd have to retrieve the associated attempt from our map before this.